src/app/CommandHandler.h

```diff
@@ -319,6 +299,54 @@ class CommandHandler
      */
     FabricIndex GetAccessingFabricIndex() const;
 
+    /**
+     * @brief Best effort to add InvokeResponse to InvokeResponseMessage.
+     *
+     * Tries to add response using lambda. Upon failure to add response, attempts
+     * to rollback the InvokeResponseMessage to a known good state. If failure is due
+     * to insufficient space in the current InvokeResponseMessage:
+     *  - Finalizes the current InvokeResponseMessage.
+     *  - Allocates a new InvokeResponseMessage.
+     *  - Reattempts to add the InvokeResponse to the new InvokeResponseMessage.
+     *
+     * addResponseFunction: A lambda function responsible for adding the response
+     *    to the current InvokeResponseMessage.
+     */
+    template <typename Function>
+    CHIP_ERROR TryAddingResponse(Function && addResponseFunction)
+    {
+        // Invalidate any existing rollback backups. The addResponseFunction is
+        // expected to create a new backup during either PrepareInvokeResponseCommand
```

Comment:
If we didn't give access to the primitives that allowed someone to muck around with the state of CommandHandler I would agree, but because we have provided access to that I think a Verify or die is riskier

Reply:
The comment though says "it is expected to create a new backup". Is there a use case where we would not? Even if this is VerifyOrLog, at least to have something that checks assumptions. I believe even if we provide people with ability to customize things we should validate invariants.

or documentation should be worded in such a way that shows that "creating a new checkpoint" is actually optional and must be handled. If no checkpoint is made it looks like we should just fail RollbackResponse() so things will handle it already.

Reply:
> The comment though says "it is expected to create a new backup". Is there a use case where we would not?

Example of how this could fail today:
1. Some primitive comes along and leaves things in an incorrect state. Because we are in a bad state, we fail to create the backup.
2. We fail to allocate a buffer, so there is nothing to back up

Reply:
Resolving: this will (correctly) return an error regardless of what the addResponseFunction does. The code is better (i.e. can rollback and fix things) if we create a backup and the comment reflects that, however this does not need to be a strongly enforced check.